plugins/basic/nexus-timeline-plugin/src/main/java/org/sonatype/nexus/timeline/feeds/subscribers/AssetSubscriber.java

```diff
@@ -0,0 +1,112 @@
+ * Eclipse Foundation. All other trademarks are the property of their respective owners.
+ */
+package org.sonatype.nexus.timeline.feeds.subscribers;
+
+import java.util.Date;
+import java.util.Map;
+
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.sonatype.nexus.common.event.Asynchronous;
+import org.sonatype.nexus.common.event.EventSubscriber;
+import org.sonatype.nexus.repository.storage.AssetCreatedEvent;
+import org.sonatype.nexus.repository.storage.AssetDeletedEvent;
+import org.sonatype.nexus.repository.storage.AssetEvent;
+import org.sonatype.nexus.repository.storage.AssetUpdatedEvent;
+import org.sonatype.nexus.repository.types.ProxyType;
+import org.sonatype.nexus.security.ClientInfo;
+import org.sonatype.nexus.timeline.feeds.FeedEvent;
+import org.sonatype.nexus.timeline.feeds.FeedRecorder;
+
+import com.google.common.collect.Maps;
+import com.google.common.eventbus.AllowConcurrentEvents;
+import com.google.common.eventbus.Subscribe;
+
+/**
+ * Subscriber listening for asset events.
+ *
+ * @since 3.0
+ */
+@Named
+@Singleton
+public class AssetSubscriber
+    extends AbstractFeedEventSubscriber
+    implements EventSubscriber, Asynchronous
+{
+  @Inject
+  public AssetSubscriber(final FeedRecorder feedRecorder) {
+    super(feedRecorder);
+  }
+
+  @Subscribe
+  @AllowConcurrentEvents
+  public void onEvent(AssetEvent e) {
+    String action;
+    if (e instanceof AssetCreatedEvent) {
+      if (ProxyType.NAME.equals(e.getRepository().getType().getValue())) {
+        action = FeedRecorder.ASSET_CACHED;
+      }
+      else {
+        action = FeedRecorder.ASSET_DEPLOYED;
+      }
+    }
+    else if (e instanceof AssetUpdatedEvent) {
+      if (ProxyType.NAME.equals(e.getRepository().getType().getValue())) {
+        action = FeedRecorder.ASSET_CACHED_UPDATE;
+      }
+      else {
+        action = FeedRecorder.ASSET_DEPLOYED_UPDATE;
+      }
+    }
+    else if (e instanceof AssetDeletedEvent) {
+      action = FeedRecorder.ASSET_DELETED;
+    }
+    else {
+      return;
+    }
+
+    final ClientInfo clientInfo = e.getClientInfo();
+
+    final Map<String, String> data = Maps.newHashMap();
+    // map is for display/templating purposes
+    putIfNotNull(data, "repoName", e.getRepository().getName());
+    putIfNotNull(data, "assetName", e.getAsset().name());
+    putIfNotNull(data, "userId", getUserId(clientInfo, "n/a"));
+    putIfNotNull(data, "userIp", clientInfo == null ? "n/a" : clientInfo.getRemoteIP());
+    putIfNotNull(data, "userUa", clientInfo == null ? "n/a" : clientInfo.getUserAgent());
+    // feed event is persisted, is searchable/filterable by these properties
+    final FeedEvent fe = new FeedEvent(
+        FeedRecorder.FAMILY_ASSET,
+        action,
+        new Date(),
+        getUserId(clientInfo, null),
+        "/repository/" + e.getRepository().getName() + e.getAsset().name(), // TODO: this is true for Maven only!
```

Comment:
this is a good reason why we should have deferred refactoring/adaption of timeline/feeds until its been thought out more.